riscv/src/elf.rs

```diff
@@ -0,0 +1,1229 @@
+) -> Vec<HighLevelInsn> {
+    try_map_two_by_two(
+        RiscVInstructionIterator::new(base_addr, data),
+        InstructionLifter {
+            rellocs_set,
+            address_map,
+            referenced_text_addrs: ReadOrWrite::Read(referenced_text_addrs),
+        },
+    )
+}
+
+struct RiscVInstructionIterator<'a> {
+    curr_address: u32,
+    remaining_data: &'a [u8],
+}
+
+impl RiscVInstructionIterator<'_> {
+    fn new(base_addr: u32, data: &[u8]) -> RiscVInstructionIterator {
+        RiscVInstructionIterator {
+            curr_address: base_addr,
+            remaining_data: data,
+        }
+    }
+}
+
+impl Iterator for RiscVInstructionIterator<'_> {
+    type Item = MaybeInstruction;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.remaining_data.is_empty() {
+            return None;
+        }
+
+        // Decide if the next instruction is 32 bits or 16 bits ("C" extension):
+        let advance;
+        let maybe_insn;
+        if self.remaining_data[0] & 0b11 == 0b11 {
+            // 32 bits
+            advance = 4;
+            let insn = u32::from_le_bytes(
+                self.remaining_data[0..4]
+                    .try_into()
+                    .expect("Not enough bytes to complete a 32-bit instruction"),
+            )
+            .decode(Isa::Rv32)
+            .unwrap_or_else(|_| {
+                // TODO: maybe instead of failing we should just emit `unimp`.
+                // This way we would support the garbage default GNU binutils
+                // linker script, that places the ELF header in the text
+                // section.
+                panic!(
+                    "Failed to decode 32-bit instruction at {:08x}",
+                    self.curr_address
+                )
+            });
+
+            maybe_insn = MaybeInstruction {
+                address: self.curr_address,
+                insn: Some(insn),
+            };
+        } else {
+            // 16 bits
+            advance = 2;
+            let bin_instruction = u16::from_le_bytes(
+                self.remaining_data[0..2]
+                    .try_into()
+                    .expect("Not enough bytes to complete a 16-bit instruction"),
+            );
+            maybe_insn = MaybeInstruction {
+                address: self.curr_address,
+                insn: match bin_instruction.decode(Isa::Rv32) {
+                    Ok(c_insn) => Some(to_32bit_equivalent(c_insn)),
+                    Err(raki::decode::DecodingError::IllegalInstruction) => {
+                        // Although not a real RISC-V instruction, sometimes 0x0000
+                        // is used on purpose as an illegal instruction (it even has
+                        // its own mnemonic "unimp"), so we support it here.
+                        // Otherwise, there is something more fishy going on, and we
+                        // panic.
+
+                        // TODO: we should just emit `unimp` for every unknown.
```

Comment:
why not done?

Reply:
It can make the generated code very big and may mask other issues. I reworded the TODO.

Reply:
Also, there are unanswered questions: should this data be readable from the program itself? What about the valid instructions?